pkg/apis/apps/v1alpha1/servicebinding_types.go

```diff
@@ -18,96 +18,92 @@ type ServiceBindingRequestSpec struct {
 	// +optional
 	EnvVarPrefix string `json:"envVarPrefix,omitempty"`
 
-	// Custom env variables
+	// DataMapping is used to map CR/CSV/CRD attributes to Custom env variables
 	// +optional
-	CustomEnvVar []corev1.EnvVar `json:"customEnvVar,omitempty"`
-
-	// BackingServiceSelector is used to identify the backing service operator.
-	// Deprecation Notice:
-	// In the upcoming release, this field would be depcreated. It would be mandatory
-	// to set "backingServiceSelectors".
-	// +optional
-	BackingServiceSelector *BackingServiceSelector `json:"backingServiceSelector,omitempty"`
+	DataMapping []corev1.EnvVar `json:"dataMapping,omitempty"`
 
 	// BackingServiceSelectors is used to identify multiple backing services.
-	// This would be made a required field after 'BackingServiceSelector'
+	// This would be made a required field after 'services'
 	// is removed.
 	// +optional
-	BackingServiceSelectors *[]BackingServiceSelector `json:"backingServiceSelectors,omitempty"`
+	Services []Service `json:"services,omitempty"`
 
-	// ApplicationSelector is used to identify the application connecting to the
+	// Application is used to identify the application connecting to the
 	// backing service operator.
 	// +optional
-	ApplicationSelector ApplicationSelector `json:"applicationSelector"`
+	Application Application `json:"application"`
 
 	// DetectBindingResources is flag used to bind all non-bindable variables from
 	// different subresources owned by backing operator CR.
 	// +optional
 	DetectBindingResources bool `json:"detectBindingResources,omitempty"`
 }
 
-// ServiceBindingRequestStatus defines the observed state of ServiceBindingRequest
+// ServiceBindingStatus defines the observed state of ServiceBindingStatus
 // +k8s:openapi-gen=true
-type ServiceBindingRequestStatus struct {
+type ServiceBindingStatus struct {
 	// Conditions describes the state of the operator's reconciliation functionality.
 	Conditions []conditionsv1.Condition `json:"conditions"`
 	// Secret is the name of the intermediate secret
-	Secret string `json:"secret"`
+	Secret corev1.LocalObjectReference `json:"secret"`
 	// ApplicationObjects contains all the application objects filtered by label
 	// +optional
 	Applications []BoundApplication `json:"applications,omitempty"`
 }
 
-// BackingServiceSelector defines the selector based on resource name, version, and resource kind
-type BackingServiceSelector struct {
-	metav1.GroupVersionKind `json:",inline"`
-	ResourceRef             string `json:"resourceRef"`
+// Service defines the selector based on resource name, version, and resource kind
+type Service struct {
+	metav1.GroupVersionKind     `json:",inline"`
```

Comment:
We can remove `metav1.GroupVersionKind` and just have `corev1.TypedLocalObjectReference`.

```
// TypedLocalObjectReference contains enough information to let you locate the
// typed referenced object inside the same namespace.
type TypedLocalObjectReference struct {
	// APIGroup is the group for the resource being referenced.
	// If APIGroup is not specified, the specified Kind must be in the core API group.
	// For any other third-party types, APIGroup is required.
	// +optional
	APIGroup *string `json:"apiGroup" protobuf:"bytes,1,opt,name=apiGroup"`
	// Kind is the type of resource being referenced
	Kind string `json:"kind" protobuf:"bytes,2,opt,name=kind"`
	// Name is the name of resource being referenced
	Name string `json:"name" protobuf:"bytes,3,opt,name=name"`
}
```

Reply:
That is an excellent idea. We could use it for `Services` only though. Since `name` is optional in `Applications` , we'll not be able to use `TypedLocalObjectReference`.

Reply:
..umm.. but `APIGroup *string` isn't supposed to accept the `Version`

Reply:
for services we need group version and Kind, `corev1.TypedLocalObjectReference` will indeed not give the Version

Reply:
Moreover corev1.TypedLocalObjectReference is for the same namespace and we can have services in different namespaces

Reply:
Perhaps [`corev1.ObjectReference`](https://pkg.go.dev/k8s.io/api/core/v1?tab=doc#ObjectReference) can be a good fit, since it has both `Namespace` and `APIVersion` members?

Regardless, can you clarify the importance of the API version in this particular case?

Reply:
I gave it a shot here https://github.com/redhat-developer/service-binding-operator/pull/495#discussion_r431446987